sound/soc/sof/sof-client.h

```diff
@@ -0,0 +1,85 @@
+/* SPDX-License-Identifier: (GPL-2.0-only) */
+
+#ifndef __SOUND_SOC_SOF_CLIENT_H
+#define __SOUND_SOC_SOF_CLIENT_H
+
+#include <linux/completion.h>
+#include <linux/debugfs.h>
+#include <linux/device.h>
+#include <linux/device/driver.h>
+#include <linux/kernel.h>
+#include <linux/list.h>
+#include <linux/virtual_bus.h>
+
+#define SOF_CLIENT_PROBE_TIMEOUT_MS 2000
+
+struct snd_sof_dev;
+
+enum sof_client_type {
+	SOF_CLIENT_AUDIO,
+	SOF_CLIENT_IPC,
+};
+
+/* SOF client device */
+struct sof_client_dev {
+	struct virtbus_device vdev;
+	struct snd_sof_dev *sdev;
+	struct list_head list;	/* item in SOF core client drv list */
+	struct completion probe_complete;
+	void *data;
+};
+
+struct sof_client_drv {
+	const char *name;
+	enum sof_client_type type;
+	const struct sof_client_ops *ops;
+	struct virtbus_driver virtbus_drv;
+};
+
+/* client-specific ops, all optional */
+struct sof_client_ops {
+	/* check if client is requesting DSP to remain in D0 during suspend */
+	bool (*pm_remain_in_D0)(struct sof_client_dev *cdev);
```

Comment:
I am now blanking on this one. Is this to keep the parent in D0 when the client goes into pm_runtime suspend, or to leave the parent in D0ix during system suspend? Either way, you've said too much or too little.